Car collision visualization c img

ahmed-nad opened the pull request.

Israr4 committed:
car collision visualization

Israr4 committed:
configuration file removed

ahmed-nad commented on Car_Collision_Visulaization/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+project(DataVisualization VERSION 0.1.0 LANGUAGES C CXX)
+
+
+add_library(
+    visualizer
+    src/CImg.h 
+    src/data_visualization.h
+    src/data_visualization.cpp
+    src/car_collision.h
+    src/car_collision.cpp
+)
```
1) Library names should start with capital letters
2) Visualizer library should only contain visualization code (car_collision.h and car_collision.cpp) should not be included as they are not part of the same functionality

So Visulization library should look like
add_library(
    Visualizer
    src/CImg.h 
    src/visualizer.h
    src/visualizer.cpp
    )

ahmed-nad commented on Car_Collision_Visulaization/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+project(DataVisualization VERSION 0.1.0 LANGUAGES C CXX)
+
+
+add_library(
```
add_library(
    Car
    src/car.h
    src/car.cpp
    )
    
    We should have meaningful names and they are separate in functionality so they should be separate libraries

ahmed-nad commented on Car_Collision_Visulaization/src/car_collision.cpp:
```diff
@@ -1,5 +1,13 @@
+
 #include "car_collision.h"
 
+    Car::Car()
+    {
+        x_        = 0;
+        y_        = 0;
+        velocity_ = 0;
+        angle_    = 0;  
+    }
 
```
We don't need this constructor please delete it

Israr4 replied:
I need that default constructor to be used in code,if i don,t write a default constructor then compiler gives error.

ahmed-nad commented on Car_Collision_Visulaization/src/car_collision.h:
```diff
@@ -9,8 +12,11 @@ class Car
     public:
     double x_, y_ ;         
     double velocity_,angle_;
-
+    Car();
     explicit Car(double, double, double, double);
+    void move();
```
please define variables and functions separately
public: 
  double x_;
  double y_;
  double velocity_;
  
public:
  explicit Car (double x, double y, double velocity, double angle);
  void move();
  bool isColliding(const Car& target_vehicle);

ahmed-nad replied:
Car should have length and width as well

ahmed-nad commented on Car_Collision_Visulaization/src/car_collision.h:
```diff
@@ -9,8 +12,11 @@ class Car
 
-bool isCalculatingCollision(const Car&, const Car&);
\ No newline at end of file
+bool isCalculatingCollision(const Car&, const Car&);
```
This should be is_colliding() what you have written does not make any sense

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.cpp:
```diff
@@ -0,0 +1,49 @@
+#include "data_visualization.h"
+
+
+Visualizer :: Visualizer()
+{
+    length_ = 0;
+    width_  = 0;
+}
```
Delete this one. And make the other one explicit  as we can not have a visualizer without a window size

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.cpp:
```diff
@@ -0,0 +1,49 @@
+    {
+        display.wait();
+    }
+}
```
This is wrong design. This way you will only get the initial values of cars and than you won't be able to do anything else. 
1) The constructor of Visualizer should have these lines
   
CImg<unsigned char> canvas(length, width, 1, 3, 255); 
CImgDisplay display(canvas, "Rectangle and Circle"); 

2) update() should accept const std::vector<Car>& cars as we want be able to draw multiple cars.
3) Car should contain length and width of car as well. Which would be used in update() to draw vehicles.

void Visualizer :: update( const std::vector<Car>& all_cars)
{
   for (const Car& car : all_cars) 
   {
       drawRectangleShape(canvas, car, red);
    }
}

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.h:
```diff
@@ -0,0 +1,26 @@
+    size_t length_;
+    size_t width_;
+    Visualizer();
+    Visualizer(size_t length, size_t width);
```
please update like i showed you in the other class

ahmed-nad commented on Car_Collision_Visulaization/main.cpp:
```diff
@@ -0,0 +1,21 @@
+    // Car car2(5, 4, 7, 135); 
+    // Visualizer visualize;
+
+    visualize.update(car1,car2); 
```
Initialize all cars in a vector .

std::vector<Car> all_cars;
all_cars.push_back(200,2,5,45);
all_cars.push_back(55,4,7,135);

while (true)
{ 
  for (const Car& car : all_cars)
    car.move();
   visualize.update(all_cars)
}

Israr4 replied:
visualize object needs default constructor there

ahmed-nad reviewed (commented):
A lot of improvements required

Israr4 committed:
New code added

Israr4 committed:
mistakes resolved

Israr4 committed:
correction

ahmed-nad commented on Car_Collision_Visulaization/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+    src/data_visualization.cpp
+)
+
+add_library(
+    Car
+    src/car_collision.h
+    src/car_collision.cpp
+)
```
car_collision.h and car_collision.cpp
are not meaningful names anymore please change it. As i suggested before please change it to car.h and car.cpp

ahmed-nad commented on Car_Collision_Visulaization/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+add_library(
+    Visualizer
+    src/CImg.h 
+    src/data_visualization.h
+    src/data_visualization.cpp
```
data_visualization.h and data_visualization.cpp are not meaningful names please change them 
as i suggested before should be 
visualizer.h and visualizer.cpp

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.h:
```diff
@@ -0,0 +1,24 @@
+
+class Visualizer 
+{
+    public:
+    size_t length_;
+    size_t width_;
+    Visualizer();
+
+    public:
+    explicit Visualizer(size_t length, size_t width);
+    void drawRectangleShape(Car car, unsigned char color[3]);
+    void update(std::vector<Car> &all_cars);
```
delete Visualizer() please we shouldn't need it

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.cpp:
```diff
@@ -0,0 +1,49 @@
+    CImg<unsigned char> canvas(length, width, 1, 3, 255); 
+    CImgDisplay display(canvas, "Rectangle and Circle"); 
+
+    for (int i = 0; i < 10; ++i) 
+    {
+        std::cout<<"Hello"<<std::endl;
+        canvas.fill(100);
+        car.move();
+        canvas.draw_rectangle(car.x_ - (carWidth / 2), car.y_ - (carHeight / 2), car.x_ + (carWidth / 2), car.y_ + (carHeight / 2), color, 1.0f);
+        canvas.display(display);
+        cimg::wait(100);
+    }
```
Why the loop here?

ahmed-nad replied:
Why put the for loop here? we don't need it.

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.cpp:
```diff
@@ -0,0 +1,49 @@
+
+void Visualizer :: drawRectangleShape( Car car, unsigned char color[3]) 
+{
+    int carWidth  = 50;
+    int carHeight = 30;
```
This should be inside each car object

ahmed-nad commented on Car_Collision_Visulaization/src/data_visualization.cpp:
```diff
@@ -0,0 +1,49 @@
+    int width  = 1000;
+    Visualizer(length,width);
+
+    CImg<unsigned char> canvas(length, width, 1, 3, 255); 
+    CImgDisplay display(canvas, "Rectangle and Circle"); 
```
This should be initialized once in the constructor

ahmed-nad reviewed (commented):
Ye design theek nai hai. Ye socho ke hr class ne kea krna hai.
1) Car ki jo class hai usmein uski length aur width honi chaiye
2) Visualizer ki class sirf canvas aur baki cheezo ko initialize kregi aur uske update() function mein jo kuch ayega usko wo draw krdegi. While loop uske andhr nai hona chaiye. Jaha se call horha while loop waha hona chaiye. Jese meine pehle bola tha ke Canvas object tm visualizer mein initialize kro jo length, width ati constructor mein uske through. Phr us canvas object ko update() function mein reuse kro.

Israr4 committed:
Indexed based loop change into range based for loop

Israr4 committed:
Description added.

Car_Collision_Visualization Branch contains C++ code to determine collisions between cars and also visualize these collision on CImg library.

Israr4 committed:
updated code

Israr4 committed:
Merge branch 'Car_Collision_Visualization_CImg' of https://github.com/Israr4/C-Tasks into Car_Collision_Visualization_CImg

Israr4 committed:
code updated

ahmed-nad commented on Car_Collision_Visulaization/src/visualizer.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void Visualizer :: drawRectangleShape(CImg<unsigned char>& canvas, Car car, unsigned char color[3])  
+{
+    std::cout<<"Hello"<<std::endl;
```
remove

Israr4 replied:
removed

Israr4 committed:
code updates

Israr4 committed:
Code Updated...
